sermant-plugins/sermant-flowcontrol/flowcontrol-plugin/src/main/java/io/sermant/flowcontrol/retry/client/HttpClient4xInterceptor.java

```diff
@@ -0,0 +1,228 @@
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+package io.sermant.flowcontrol.retry.client;
+
+import io.github.resilience4j.retry.Retry;
+import io.sermant.core.common.LoggerFactory;
+import io.sermant.core.plugin.agent.entity.ExecuteContext;
+import io.sermant.core.service.xds.entity.ServiceInstance;
+import io.sermant.flowcontrol.AbstractXdsHttpClientInterceptor;
+import io.sermant.flowcontrol.common.config.CommonConst;
+import io.sermant.flowcontrol.common.entity.FlowControlResult;
+import io.sermant.flowcontrol.common.entity.FlowControlScenario;
+import io.sermant.flowcontrol.common.entity.HttpRequestEntity;
+import io.sermant.flowcontrol.common.entity.RequestEntity;
+import io.sermant.flowcontrol.common.handler.retry.AbstractRetry;
+import io.sermant.flowcontrol.common.handler.retry.RetryContext;
+import io.sermant.flowcontrol.common.util.XdsRouterUtils;
+import io.sermant.flowcontrol.common.util.XdsThreadLocalUtil;
+import io.sermant.flowcontrol.entity.ErrorCloseableHttpResponse;
+
+import org.apache.http.Header;
+import org.apache.http.HttpHost;
+import org.apache.http.HttpRequest;
+import org.apache.http.HttpStatus;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpRequestBase;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * HTTP interception only for version 4. x
+ *
+ * @author zhp
+ * @since 2024-12-20
+ */
+public class HttpClient4xInterceptor extends AbstractXdsHttpClientInterceptor {
+    private static final Logger LOGGER = LoggerFactory.getLogger();
+
+    private final String className = HttpClient4xInterceptor.class.getName();
+
+    private final io.sermant.flowcontrol.common.handler.retry.Retry retry = new HttpClientRetry();
+
+    /**
+     * Pre-trigger point
+     *
+     * @param context Execution context
+     * @return Execution context
+     */
+    @Override
+    public ExecuteContext doBefore(ExecuteContext context) {
+        Object[] arguments = context.getArguments();
+        Object httpRequestObject = arguments[1];
+        if (!(httpRequestObject instanceof HttpRequestBase)) {
+            return context;
+        }
+        final HttpRequestBase httpRequest = (HttpRequestBase) httpRequestObject;
+        if (!needHandleXdsRouter(httpRequest) && RetryContext.INSTANCE.getRetryPolicy() == null) {
```

Comment:
move this judgement to convertToHttpEntity

Reply:
done